Get prelude.kore from the backend Nix expression

ttuegel opened the pull request:
This allows to override the `prelude.kore` used in the Nix expression without
any Git submodule trickery.

radumereuta commented:
@bmmoore please review.
I've been working with @cos to fix issue #1705.
This replaces an older try of mine which wasn't really clean.
Basically we've extended the implementation of Module to automatically add the subsort production between two lists which have related elements. This is done automatically, and is consistent with what @grosu was saying in #1733.

The only problem with doing this automatically is that we don't want that production to be included in the parser for programs, and I solve this by a simple check when I'm generating the grammar for programs.

This should also solve some recent issues with Simple.

bmmoore commented:
The line I commented on looks like a mistake, please explain it. I didn't see a compensating change elsewhere, so I don't know how both versions can even pass tests, unless that attribute is unused or we have insufficient tests.

Have you updated any definitions to depend on the extra subsorting? It seems like this shouldn't affect parsing at all, is it useful for anything yet?

My commit should be a simple merge to catch up with changes in master, but please check over it. There was a merge conflict in `outer.scala`.

cos commented:
@bmmoore, yes, that bug slipped in, thanks!

@radumereuta, if you know where such sub-sorting appears in the tutorial, could you have a go at it? Otherwise, I can make the changes.

radumereuta commented:
The test that I added doesn't work without the inferred subsort.
In the tutorial, this is needed at least for Simple. (see #1827 @daejunpark comment from Nov 10).

bmmoore commented:
The fix looks fine.

cos commented:
correct

radumereuta commented:
So, ok to be merged now?

bmmoore commented:
Add a test that would have caught that bug.

radumereuta commented:
Which bug? The subsort to KResult? That should have been taken care by the `kool` programs.
But what baffles me right now, is that the test failed, then I merged master and they passed, although they shouldn't, then I've re-added the subsort, and passed again.

bmmoore commented:
I mean the one fixed in 8a85bec, where you had been giving the nil production the same klabel as the cons productions.

radumereuta commented:
Does this test look good?

bmmoore commented:
That looks good, as long as actually it fails if you (temporarily) undo 8a85bec.

bmmoore commented:
That test doesn't work, it's the `Es#Terminator ::= "" [klabel('.Es)]` production which was affected.

radumereuta commented:
Yes, you are right, I forgot to test for programs also.

bmmoore commented:
That test also doesn't find any problems. You should be able to run a command like this and get the last message:

    $ mvn package && git revert --no-edit 8a85bec && ! mvn package && echo "Test is strong enough"

Actually, the apply function in `definition.Production` adds an additional klabel attribute based on the first argument, so some code might work after reverting 8a85bec depending on the set order. But, your test still "passes" even if after reverting that commit I also edit RuleGrammarGenerator.java#274 to construct the production as

    prod1 = Production(ul.klabel, Sort(ul.sort + "#Terminator"), Seq(Terminal("")),
            newAtts.add(Constants.ORIGINAL_PRD, ul.pTerminator))

We need a regression test that would actually test that these productions are actually used correctly in the parser, or perhaps to explain why the klabel assigned there is never actually used.

radumereuta commented:
Right again.
I made sure this time that all the productions generated for program lists are verified in the test, for both `*` and `+`.

cos commented:
@radumereuta, this is more your PR than mine. 

If you consider the issue fixed, please mark it as `ready-for-review` so @bmmoore can merge it.

bmmoore commented:
Okay, these tests still don't actually catch the bug fixed in 8a85bec, but everything else seems to work (and we've fixed that bug), so I'm going to merge this and open a new pull request for sorting out the regression tests.

bmmoore commented:
I made pull request #1917, which starts by reverting 8a85bec, to continue figuring out why those tests don't work.

ttuegel committed:
Get prelude.kore from the backend Nix expression

This allows to override the `prelude.kore` used in the Nix expression without
any Git submodule trickery.

ttuegel committed:
Merge branch 'master' into nix-prelude-kore

ttuegel committed:
Merge branch 'master' into nix-prelude-kore

rv-jenkins committed:
Merge branch 'master' into nix-prelude-kore
